spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java

```diff
@@ -109,88 +112,167 @@ public void sawOpcode(int seen) {
-                                .addClassAndMethod(this)
-                                .addReferencedField(this)
-                                .add(LocalVariableAnnotation.getLocalVariableAnnotation(getMethod(), top.getRegisterNumber(),
-                                        getPC(), getPC() - 1)), this);
+                        new BugInstance(this, "EI_EXPOSE_STATIC_" + (capture == CaptureKind.BUF ? "BUF2" : "REP2"),
+                                capture == CaptureKind.BUF ? LOW_PRIORITY : NORMAL_PRIORITY)
+                                        .addClassAndMethod(this)
+                                        .addReferencedField(this)
+                                        .add(LocalVariableAnnotation.getLocalVariableAnnotation(getMethod(),
+                                                top.getRegisterNumber(),
+                                                getPC(), getPC() - 1)), this);
             }
         }
         if (!staticMethod && seen == Const.PUTFIELD && nonPublicFieldOperand()
                 && MutableClasses.mutableSignature(getSigConstantOperand())) {
             OpcodeStack.Item top = stack.getStackItem(0);
             OpcodeStack.Item target = stack.getStackItem(1);
-            if (isPotentialCapture(top) && target.getRegisterNumber() == 0) {
+            CaptureKind capture = getPotentialCapture(top);
+            if (capture != CaptureKind.NONE && target.getRegisterNumber() == 0) {
                 bugAccumulator.accumulateBug(
-                        new BugInstance(this, "EI_EXPOSE_REP2", NORMAL_PRIORITY)
-                                .addClassAndMethod(this)
-                                .addReferencedField(this)
-                                .add(LocalVariableAnnotation.getLocalVariableAnnotation(getMethod(), top.getRegisterNumber(),
-                                        getPC(), getPC() - 1)), this);
+                        new BugInstance(this, "EI_EXPOSE_" + (capture == CaptureKind.BUF ? "BUF2" : "REP2"),
+                                capture == CaptureKind.BUF ? LOW_PRIORITY : NORMAL_PRIORITY)
+                                        .addClassAndMethod(this)
+                                        .addReferencedField(this)
+                                        .add(LocalVariableAnnotation.getLocalVariableAnnotation(getMethod(),
+                                                top.getRegisterNumber(),
+                                                getPC(), getPC() - 1)), this);
             }
         }
 
-        if (seen == Const.ALOAD_0 && !staticMethod) {
-            thisOnTOS = true;
-            fieldOnTOS = false;
-            return;
+        if (seen == Const.ARETURN) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            XField field = item.getXField();
+            boolean isBuf = false;
+            if (field == null) {
+                field = bufferFieldDuplicates.get(item);
+                isBuf = true;
```

Comment:
This line will execute even if a buffer wasn't actually found, which seems error-prone.